src/decision-engine/index.js

```diff
@@ -49,15 +49,26 @@ class DecisionEngine {
 
     let size = 0
     let batch = []
+    let outstanding = blocks.length
 
     eachSeries(blocks, (b, cb) => {
+      outstanding--
       batch.push(b)
       size += b.data.byteLength
 
-      if (size >= MAX_MESSAGE_SIZE) {
+      if (size >= MAX_MESSAGE_SIZE ||
+          // need to ensure the last remaining items get sent
+          outstanding === 0) {
         const nextBatch = batch.slice()
         batch = []
-        this._sendSafeBlocks(env.peer, nextBatch, cb)
+        this._sendSafeBlocks(env.peer, nextBatch, (err) => {
```

Comment:
Just noticed now that we have this `env` thing. I'm curious to learn why have you adopted this strategy, doesn't seem to match your love for things like Flow where every argument gets a type and every function a signature. Here it is impossible for the reader to figure out the amount of variables that can go into this `env` object.